optuna/integration/cma.py

```diff
@@ -0,0 +1,438 @@
+
+        self._search_space = search_space
+        self._param_names = list(sorted(self._search_space.keys()))
+
+        lows = []
+        highs = []
+        for param_name in self._param_names:
+            dist = self._search_space[param_name]
+            if isinstance(dist, CategoricalDistribution):
+                # Handle categorical values by ordinal representation.
+                # TODO(Yanase): Support one-hot representation.
+                lows.append(-0.5)
+                highs.append(len(dist.choices) - 0.5)
+            elif isinstance(dist, UniformDistribution) or \
+                    isinstance(dist, LogUniformDistribution):
+                lows.append(self._to_cma_params(search_space, param_name, dist.low))
+                highs.append(self._to_cma_params(search_space, param_name, dist.high))
+            elif isinstance(dist, DiscreteUniformDistribution):
+                r = dist.high - dist.low
+                lows.append(0 - 0.5 * dist.q)
+                highs.append(r + 0.5 * dist.q)
+            elif isinstance(dist, IntUniformDistribution):
+                lows.append(dist.low - 0.5)
+                highs.append(dist.high + 0.5)
+            else:
+                raise NotImplementedError('The distribution {} is not implemented.'.format(dist))
+
+        # Set initial params.
+        initial_cma_params = []
+        for param_name in self._param_names:
+            initial_cma_params.append(
+                self._to_cma_params(self._search_space, param_name, x0[param_name]))
+        cma_option = {
+            'BoundaryHandler': cma.BoundTransform,
+            'bounds': [lows, highs],
+        }
+
+        if cma_stds:
+            cma_option['CMA_stds'] = [cma_stds.get(name, 1.) for name in self._param_names]
+
+        cma_opts.update(cma_option)
+
+        self._es = cma.CMAEvolutionStrategy(initial_cma_params, sigma0, cma_opts)
+
+    def tell(self, trials, study_direction):
+        # type: (List[FrozenTrial], StudyDirection) -> int
+
+        complete_trials = []
+        for trial in trials:
+            if trial.distributions != self._search_space:
+                continue
+            if trial.state != TrialState.COMPLETE:
+                continue
+            if not self._is_compatible(trial):
+                continue
+            complete_trials.append(trial)
+
+        popsize = self._es.popsize
+        generation = len(complete_trials) // popsize
+        for i in range(generation):
+            xs = []
+            ys = []
+            for t in complete_trials[i * popsize:(i + 1) * popsize]:
+                x = [
+                    self._to_cma_params(self._search_space, name, t.params[name])
+                    for name in self._param_names
+                ]
+                xs.append(x)
+                ys.append(t.value)
+            if study_direction == StudyDirection.MAXIMIZE:
+                ys = [-1 * y if y is not None else y for y in ys]
+            self._es.ask()
```

Comment:
Could you add a brief comment about why this `ask()` call is needed?
(for the future readers)

Reply:
That sounds good.
Fixed: https://github.com/pfnet/optuna/pull/447/commits/2c5f0a1c0db21c2908788c9c540afde33236be21

Reply:
Thank you!